src/rgw/motr/gc/gc_sync.cc

```diff
@@ -0,0 +1,133 @@
+ *
+ */
+
+#include "motr/gc/gc_sync.h"
+
+int MotrGCThreadSync::lock(const std::string &lock_name, MotrLockType lock_type,
+                           utime_t lock_duration, const std::string locker_id) {
+  if (!motr_gclock || locker_id.empty()) {
+    return -EINVAL;
+  }
+  int rc = 0;
+  motr_lock_info_t lock_obj;
+  rc = motr_gclock->read_lock(lock_name, &lock_obj);
+  if (rc == -ENOENT) {
+    // lock is available/free
+  } else {
+    // lock entry is present. Check if this is a stale/expired lock
+    utime_t now = ceph_clock_now();
+    auto iter = lock_obj.lockers.begin();
+    while (iter != lock_obj.lockers.end()) {
+      struct motr_locker_info_t &info = iter->second;
+      if (!info.expiration.is_zero() && info.expiration < now) {
+        // locker has expired; delete it
+        iter = lock_obj.lockers.erase(iter);
+      } else {
+        ++iter;
+      }
+    }
+    if (!lock_obj.lockers.empty()) {
+      // lock is not available
+      return -EBUSY;
+    } else {
+      // Try to acquire lock object
+      struct motr_locker_info_t locker;
+      locker.coockie = locker_id;
+      locker.expiration = ceph_clock_now() + lock_duration;
+      locker.description = "";
+      // Update lock entry with current locker and lock the resource
+      lock_obj.lockers.insert(
+          std::pair<std::string, struct motr_locker_info_t>(locker_id,locker));
+      rc = motr_gclock->write_lock(lock_name, &lock_obj, false);
+      if (rc < 0) {
+        // Failed to acquire lock object; possibly, already acquired
+      }
+      return rc;
+    }
+  }
+
+  return 0;
+}
+
+int MotrGCThreadSync::unlock(const std::string &lock_name,
+                             MotrLockType lock_type,
+                             const std::string locker_id) {
+  return motr_gclock->remove_lock(lock_name, locker_id);
+}
+
+void MotrKVLockProvider::initialize(const DoutPrefixProvider* dpp,
+                                    rgw::sal::MotrStore* _s,
+                                    std::string& lock_index_name) {
+  _dpp = dpp;
+  _store = _s;
+  _lock_index = lock_index_name;
+}
+
+int MotrKVLockProvider::read_lock(const std::string &lock_name,
+                                  motr_lock_info_t* lock_info) {
+  if (!_store || _lock_index.empty() || !lock_info) {
+    return -EINVAL;
+  }
+  int rc = 0;
+  bufferlist bl;
+  rc = _store->do_idx_op_by_name(_lock_index, M0_IC_GET, lock_name, bl);
+  if (rc != 0) {
+    return rc;
+  }
+  bufferlist::const_iterator bitr = bl.begin();
+  lock_info->decode(bitr);
+  return 0;
+}
+
+int MotrKVLockProvider::write_lock(const std::string &lock_name,
+                                   motr_lock_info_t* lock_info, bool update) {
+  if (!_store || _lock_index.empty() || !lock_info) {
+    return -EINVAL;
+  }
+  bufferlist bl;
+  lock_info->encode(bl);
+  int rc =
+      _store->do_idx_op_by_name(_lock_index, M0_IC_PUT, lock_name, bl, update);
+  if (rc != 0) {
+    return rc;
+  }
+  return 0;
```

Comment:
```suggestion
  return rc;
```

Reply:
yes..as nothing is done in code when rc!=0, so better to return the rc.